dll/win32/imm32/imm.c

```diff
@@ -4869,47 +4870,80 @@ BOOL WINAPI ImmTranslateMessage(HWND hwnd, UINT msg, WPARAM wParam, LPARAM lKeyD
 *		ImmProcessKey(IMM32.@)
 *       ( Undocumented, called from user32.dll )
 */
-BOOL WINAPI ImmProcessKey(HWND hwnd, HKL hKL, UINT vKey, LPARAM lKeyData, DWORD unknown)
+BOOL WINAPI ImmProcessKey(HWND hWnd, HKL hKL, UINT vKey, LPARAM lParam, DWORD dwHotKeyID)
 {
-    InputContextData *data;
-    HIMC imc = ImmGetContext(hwnd);
-    BYTE state[256];
-
-    TRACE("%p %p %x %x %x\n",hwnd, hKL, vKey, (UINT)lKeyData, unknown);
+    DWORD ret = 0;
+    HIMC hIMC;
+    PIMEDPI pImeDpi;
+    LPINPUTCONTEXTDX pIC;
+    BYTE KeyState[256];
+    UINT vk;
+    BOOL bUseIme = TRUE, bSkipThisKey = FALSE, bCut = FALSE;
 
-    if (imc)
-        data = (InputContextData *)imc;
-    else
-        return FALSE;
+    TRACE("(%p, %p, 0x%X, %p, 0x%lX)\n", hWnd, hKL, vKey, lParam, dwHotKeyID);
 
-    /* Make sure we are inputting to the correct keyboard */
-    if (data->immKbd->hkl != hKL)
+    hIMC = ImmGetContext(hWnd);
+    pImeDpi = ImmLockImeDpi(hKL);
+    if (pImeDpi)
     {
-        ImmHkl *new_hkl = IMM_GetImmHkl(hKL);
-        if (new_hkl)
+        pIC = (LPINPUTCONTEXTDX)ImmLockIMC(hIMC);
+        if (pIC)
         {
-            data->immKbd->pImeSelect(imc, FALSE);
-            data->immKbd->uSelected--;
-            data->immKbd = new_hkl;
-            data->immKbd->pImeSelect(imc, TRUE);
-            data->immKbd->uSelected++;
+            if (LOBYTE(vKey) == VK_PACKET &&
+                !(pImeDpi->ImeInfo.fdwProperty & IME_PROP_ACCEPT_WIDE_VKEY))
+            {
+                if (pImeDpi->ImeInfo.fdwProperty & IME_PROP_UNICODE)
+                {
+                    bCut = TRUE;
+                }
+                else
+                {
+                    bUseIme = FALSE;
+                    if (pIC->fOpen)
+                        bSkipThisKey = TRUE;
+                }
+            }
+
+            if (bUseIme)
+            {
+                if (GetKeyboardState(KeyState))
+                {
+                    vk = (bCut ? LOWORD(vKey) : vKey);
```

Comment:
"cut" ? meaning of this variable is not clear at all.

Reply:
Renamed as bLowWordOnly.